### PR TITLE
Migrate away from explicit comparisons of GeomDetEnumerators::PixelBarrel/PixelEndcap

### PR DESCRIPTION
**Pull request by makortel**
This PR migrates all code (found with `git grep GeomDetEnumerators::Pixel`) from using `GeomDetEnumerators::PixelBarrel/PixelEndcap` to either use the helper functions in GeomDetEnumerators, or `GeomDetEnumerators::subDetGeom` (one case) (one case was also from `subDetGeom` to the helper functions). These changes are needed for the Phase1/2 upgrades, where pixel barrel has `P1PXB` and pixel endcap `P1PXEC` or `P2PXEC` values. (Something similar is probably needed for Phase2 strips, but that is left for future work).

I also added a simple unit test for GeomDetEnumerators to ensure that all subdets are either `isBarrel()` or `isEndcap()`, so that I can safely assume in production code that, according to these functions, if something is not barrel, it is endcap (and vice versa). (Mostly because the condition that spurred this PR was `if(pixel barrel) ... else if(pixel endcap) ... else throw`).

Together with #11621 and #11609, the reco step of phase1 workflow `runTheMatrix.py --what upgrade -l 10000` technically works. (The final HARVESTING step fails on missing input file, because DQM/VALIDATION are not part of the reco step)

Tested in 760pre6, no changes expected in Run1/2 workflows.

@rovere @VinInn @boudoul @venturia @mark-grimes

### CODE REVIEW

**Comment by diguida** [comment]
please test

**Comment by diguida** [comment]
+1

**Comment by lveldere** [comment]
+1

**Comment by ianna** [comment]
+1

File: RecoTracker/TkNavigation/test/NavigationSchoolAnalyzer.cc [hunk 1]
```diff
@@ -75,36 +75,26 @@ void NavigationSchoolAnalyzer::print(std::ostream& os,const DetLayer* dl){
   unsigned int LorW=0;
   unsigned int side=0;
 
+  if(GeomDetEnumerators::isTracker(dl->subDetector())) {
+    LorW = tTopo->layer(tag->geographicalId());
+    side = tTopo->side(tag->geographicalId());
+  }
+  else {
   switch (dl->subDetector()){
```
Comment (cvuosalo) [inline RecoTracker/TkNavigation/test/NavigationSchoolAnalyzer.cc]: It looks like you accidentally deleted the entry for drift tubes:

 case GeomDetEnumerators::DT :
    LorW = DTChamberId(tag->geographicalId().rawId()).station(); break;

Aren't these lines needed?

File: RecoTracker/TkTrackingRegions/src/RectangularEtaPhiTrackingRegion.cc [hunk 2]
```diff
@@ -324,10 +323,12 @@ TrackingRegion::Hits RectangularEtaPhiTrackingRegion::hits(
     const GlobalPoint vtx = origin();
     GlobalVector dir = direction();
     
-    if (detLayer->subDetector() == GeomDetEnumerators::PixelBarrel || (!theUseEtaPhi  && detLayer->location() == GeomDetEnumerators::barrel)){
+    if ((GeomDetEnumerators::isTrackerPixel(detLayer->subDetector()) && GeomDetEnumerators::isBarrel(detLayer->subDetector()) == GeomDetEnumerators::PixelBarrel) ||
+        (!theUseEtaPhi  && detLayer->location() == GeomDetEnumerators::barrel)) {
       const BarrelDetLayer& bl = dynamic_cast<const BarrelDetLayer&>(*detLayer);
       est = estimator(&bl,es);
-    } else if (detLayer->subDetector() == GeomDetEnumerators::PixelEndcap || (!theUseEtaPhi  && detLayer->location() == GeomDetEnumerators::endcap)) {
+    } else if ((GeomDetEnumerators::isTrackerPixel(detLayer->subDetector()) && GeomDetEnumerators::isBarrel(detLayer->subDetector()) == GeomDetEnumerators::PixelEndcap) ||
```
Comment (cvuosalo) [inline RecoTracker/TkTrackingRegions/src/RectangularEtaPhiTrackingRegion.cc]: isBarrel() returns a bool. Comparing it to GeomDetEnumerators::PixelEndcap looks incorrect. Same comment for line 326.

File: RecoTracker/TkNavigation/test/NavigationSchoolAnalyzer.cc (hunk 1 above)
Comment (makortel) [inline RecoTracker/TkNavigation/test/NavigationSchoolAnalyzer.cc]: Whoops, yes, thanks. The intention was to remove only the tracker cases from the switch.

File: RecoTracker/TkTrackingRegions/src/RectangularEtaPhiTrackingRegion.cc (hunk 2 above)
Comment (makortel) [inline RecoTracker/TkTrackingRegions/src/RectangularEtaPhiTrackingRegion.cc]: Yes, those should be removed.

File: RecoMuon/TrackerSeedGenerator/plugins/TSGForRoadSearch.cc [hunk 3]
```diff
@@ -160,7 +160,7 @@ void TSGForRoadSearch::makeSeeds_0(const reco::Track & muon, std::vector<Traject
 
   //loop the parts until at least a compatible is found
   while (compatible.size()==0) {
-    switch ( inLayer->subDetector() ) {
+    switch ( GeomDetEnumerators::subDetGeom[inLayer->subDetector()] ) {
```
Comment (makortel) [inline RecoMuon/TrackerSeedGenerator/plugins/TSGForRoadSearch.cc]: While we're at it, after some private e-mails I'm no longer sure if this (use of `subDetGeom[]`) is the best solution. Although it solves the problem for Phase1 (`P1PXB` and `P1PXEC` get mapped to `PixelBarrel` and `PixelEndcap`), it also follows that the Phase2 enums get mapped by `P2PXEC -> PixelEndcap`, `P2OTB->TOB`, `P2OTEC->TID` which may be fine here or not (I have no idea). Alternative would be to add `P1PXB` and `P1PXEC` cases here, and leave the Phase2 cases to be dealt later (the two other switches in this file use LogError, so there is some hope to get notified from that).

I'd like to hear second opinion(s), and I'll leave the change as it is for now.

**Comment by makortel** [comment]
Addressed @cvuosalo's comments

**Comment by cvuosalo** [comment]
@cmsbuild please test

**Comment by cvuosalo** [comment]
+1

For #11624 902fd4991a40a368050a9ce670dd4aa457f0cd22

Converting detector comparisons of the pixel detector to support Phase 1 changes. There should be no change in the logic, only the method of comparison. There should be no change in run-time behavior.

The code changes are satisfactory, and Jenkins [tests](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-11624/8654/summary.html) against baseline CMSSW_7_6_X_2015-10-05-2300 show no significant differences, as expected.

**Comment by ianna** [comment]
+1

**Comment by lveldere** [comment]
+1

**Comment by makortel** [comment]
@diguida @mmusich Could you please sign again? Thanks.

**Comment by diguida** [comment]
+1
Concerning your comment https://github.com/cms-sw/cmssw/pull/11624#discussion_r41305873 I hope @ianna has a better opinion.

**Comment by davidlange6** [comment]
+1

**Comment by slava77** [comment]
@makortel @VinInn 
what is this issue in  902fd49  affecting?
Apparently we didn't catch any changes in the tests.

**Comment by makortel** [comment]
Quoting Vincenzo from the private thread

> it is higly probable that the bug does not affect anybody as it requires
UseMeasurementTracker::kAlways  (or in python UseMeasurementTracker = 1 )
which does not seem very popular

So probably the damage is minimal (or even non-existent?). IIRC in the typical workflows (like in PR tests) this branch of the code is not run.

**Comment by VinInn** [comment]
Indeed, validating dead code is tought

**Pull request merged**